src/internal/permissions.ts

```diff
@@ -0,0 +1,429 @@
+export interface PermissionHandle {
+  // Question: Best way to approach time or date here. Date object, Epoch time string, time in milliseconds
+  // The contract will have this as an epoch time in seconds as a number
+  expiration: number;
+  // Question: Should this be a field at all? Should dapps be able to raise their own rate limit?
+  // Question: Should this be a number (1, 10, 100, etc) or a string (low, medium, high)?
+  transactionsPerSecond: number;
+  permissions: Permission[];
+}
+
+// Actions that can be granted on NFTs
+export enum NFTCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export enum PermissionType {
+  FungibleAsset = "FungibleAsset",
+  Gas = "Gas",
+  NFT = "NFT",
+  NFTCollection = "Collection",
+}
+
+export enum MoveVMPermissionType {
+  FungibleAsset = "0x1::fungible_asset::WithdrawPermission",
+  
+}
+
+export interface FungibleAssetPermission {
+  type: PermissionType.FungibleAsset;
+  asset: string;
+  // Question: best type here?: number | string | bigint
+  balance: string;
+}
+
+export interface GasPermission {
+  type: PermissionType.Gas;
+  amount: number;
+}
+
+export interface NFTPermission {
+  type: PermissionType.NFT;
+  assetAddress: string;
+  capabilities: Record<NFTCapability, boolean>;
+}
+
+export interface NFTCollectionPermission {
+  type: PermissionType.NFTCollection;
+  collectionAddress: string;
+  capabilities: Record<NFTCollectionCapability, boolean>;
+}
+
+export enum NFTCollectionCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export interface GenericPermission {
+  type: Exclude<string, PermissionType | `${PermissionType}`>;
+  [key: string]: string | Exclude<string, PermissionType | `${PermissionType}`>;
+}
+
+// revoke permission types
+export type RevokeFungibleAssetPermission = Pick<FungibleAssetPermission, "type" | "asset">;
+export type RevokeNFTPermission = Pick<NFTPermission, "type" | "assetAddress">;
+
+// factories -- on second thought, i think these are dumb
+export function FungibleAssetPermission(args: Omit<FungibleAssetPermission, "type">): FungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function GasPermission(args: Omit<GasPermission, "type">): GasPermission {
+  return { type: PermissionType.Gas, ...args };
+}
+export function NFTPermission(args: Omit<NFTPermission, "type">): NFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
+export function NFTCollectionPermission(args: Omit<NFTCollectionPermission, "type">): NFTCollectionPermission {
+  return { type: PermissionType.NFTCollection, ...args };
+}
+
+// revoke factories
+export function RevokeFungibleAssetPermission(args: Omit<RevokeFungibleAssetPermission, "type">): RevokeFungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function RevokeNFTPermission(args: Omit<RevokeNFTPermission, "type">): RevokeNFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
```

Comment:
The main draw here is improved type inference. To get a feel for how it would work without the factories, try consuming `requestPermissions` without using factories. You open a new object and are left without guidance on what the first step is. A brand new consumer of the API might not find it readily apparent that all the properties are defined based on what `type` is specified.

Might be good to brainstorm other possible DX around this.

Reply:
should factories be moved to api file?

Reply:
but good point actually

Reply:
Since they share the same name as the interface, I think having them both be here is fine.

Reply:
@hardsetting to give some input here

Reply:
I'm surprised that typing `{ type: 'permission1', ... }` would not provide autocompletion strictly for `Permission1`.
That's the point of discriminated unions. Could it be the IDE not working correctly? lemme try a couple ones

Reply:
This is with VSCode

![Screenshot 2024-12-04 at 12 19 16 PM](https://github.com/user-attachments/assets/8d82b311-8de3-43be-ba3c-dbc72c649bda)

Reply:
But if you want to use factories go for it, I would maybe use some different naming like
```typescript
export function makeFungibleAssetPermission(...);
export function makeGasPermission(...);
```

they're still functions after all

Reply:
what's the point of the "revoke" factories?

Reply:
also try to split the code in different files / folders.
types and factories are a good candidate for different files

Reply:
information required for revoking is a subset of what's needed for granting permission so it's just a builder for a smaller object